File: 100-Days-Series/Day 100/GraphRepresentationUsingAdjacencyList.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node {
    int data;
    struct Node* next;
} Node;
typedef struct Graph {
    int numVertices;
    Node** adjList;
} Graph;
// Function to create a new node
Node* createNode(int data) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}
// Function to create a graph
Graph* createGraph(int vertices) {
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->numVertices = vertices;
    graph->adjList = (Node**)malloc(vertices * sizeof(Node*));

    for (int i = 0; i < vertices; i++) {
        graph->adjList[i] = NULL;
    }

    return graph;
}
// Function to add an edge to the graph
void addEdge(Graph* graph, int u, int v) {
    // Add edge from u to v
    Node* newNode = createNode(v);
    newNode->next = graph->adjList[u];
    graph->adjList[u] = newNode;
    // Add edge from v to u (since the graph is undirected)
    newNode = createNode(u);
    newNode->next = graph->adjList[v];
    graph->adjList[v] = newNode;
}
// Function to print the graph
void printGraph(Graph* graph) {
    for (int v = 0; v < graph->numVertices; v++) {
        printf("\nNeighbors of Node %d: ", v + 1);
        Node* temp = graph->adjList[v];
        while (temp) {
            printf("->%d ", temp->data + 1);
            temp = temp->next;
        }
    }
    printf("\n");
}
// Let's Start Here
void main() {
    int v = 5; // Number of vertices
    Graph* graph = createGraph(v);
    // Add edges
    addEdge(graph, 0, 1); // Edge between 1 and 2
    addEdge(graph, 0, 2); // Edge between 1 and 3
    addEdge(graph, 1, 3); // Edge between 2 and 4
    addEdge(graph, 2, 3); // Edge between 3 and 4
    addEdge(graph, 2, 4); // Edge between 3 and 5
    addEdge(graph, 3, 4); // Edge between 4 and 5
    // Print the adjacency list
    printGraph(graph);
}

File: 100-Days-Series/Day 100/GraphRepresentationUsingAdjacencyList.py
class Node:
    def __init__(self, data):
        self.data = data


class Graph:
    def __init__(self, vertices):
        # Initialize adjacency list with empty lists for each vertex
        self.adj_list = [[] for _ in range(vertices)]

    def add_edge(self, u, v):
        # Add edge from u to v (undirected graph)
        self.adj_list[u].append(Node(v))
        self.adj_list[v].append(Node(u))

    def print_graph(self):
        for v in range(len(self.adj_list)):
            neighbors = " -> ".join(str(neighbor.data + 1) for neighbor in self.adj_list[v])
            print(f"Neighbors of Node {v + 1}: {neighbors}")



vertices = 5  # Number of vertices
graph = Graph(vertices)

# Add edges
graph.add_edge(0, 1)  # Edge between 1 and 2
graph.add_edge(0, 2)  # Edge between 1 and 3
graph.add_edge(1, 3)  # Edge between 2 and 4
graph.add_edge(2, 3)  # Edge between 3 and 4
graph.add_edge(2, 4)  # Edge between 3 and 5
graph.add_edge(3, 4)  # Edge between 4 and 5
# Print adjacency list
graph.print_graph()


